core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/common/iterator/AbstractRandomChangeIterator.java

```diff
@@ -37,24 +39,31 @@ protected Move_ createUpcomingSelection() {
         }
         Object entity = entityIterator.next();
 
+        VariableDescriptor<Solution_> variableDescriptor = valueSelector.getVariableDescriptor();
         Iterator<Object> valueIterator = valueSelector.iterator(entity);
         int entityIteratorCreationCount = 0;
-        // This loop is mostly only relevant when the entityIterator or valueIterator is non-random or shuffled
-        while (!valueIterator.hasNext()) {
-            // Try the next entity
-            if (!entityIterator.hasNext()) {
-                entityIterator = entitySelector.iterator();
-                entityIteratorCreationCount++;
-                if (entityIteratorCreationCount >= 2) {
-                    // All entity-value combinations have been tried (some even more than once)
-                    return noUpcomingSelection();
+        while (true) {
```

Comment:
There could be no doable moves. If that happens, it's very bad.
Some situations of no doable moves can be detected, for example with ChangeMove or SwapMove selectors that see there is only 1 planning value. But for more complex moves, not all situations of no doable moves can be detected without false positives.

If there are no doable moves, this **loops forever**. Calling solver.terminateEarly() will NOT end it. Time spend up will NOT end it. Asking the JVM to shutdown, which will interrupte its thread, will NOT end it. It will need to be killed with fire (kill -9). That kills OptaPlanner's reputation.

The move evaluation loop does check termination (by terminateEarly, by interruption or by configured termination such as timeSpent), so users think optaplanner stops immediately if the time is up, which is important for them. In fact, optaplanner stops at the first next move (or next step or next phase) iteration, which is as good as immediately normally.

@triceo @yurloc I found it good to be paranoid about avoiding "forever loops".